### ISSUE TITLE
email preview without deploying

### ISSUE THREAD

**Issue by aronatkins** [rich-iannone/quarto-email#1]
It should be possible to preview the email message without deploying and sending email.

I have been using this R function to see the contents of the generated message without deploying, which makes iteration during development much more straight-forward. It lets me see the HTML with the embedded images, which is enough in most cases.

```r
email_preview <- function(dir = getwd()) {
  email_dir <- file.path(dir, "email-preview")
  dir.create(email_dir, showWarnings = FALSE)
  
  json_file <- file.path(dir, ".output_metadata.json")
  payload <- jsonlite::fromJSON(json_file)
  
  html <- payload$rsc_email_body_html
  html <- gsub('src="cid:', 'src="', html)
  html_file <- file.path(email_dir, "index.html")
  writeLines(html, html_file)
  
  images <- payload$rsc_email_images
  for (image in names(images)) {
    encoded <- images[image]
    decoded <- base64enc::base64decode(paste0(encoded, collapse = "\n"))
    writeBin(decoded, file.path(email_dir, image))
  }
  
  html_file
}  
```

From the R console, after rendering my document, I run:

```r
browseURL(email_preview())
```

**Comment by aronatkins** [comment]
A similar helper in Python:

```python
#!/usr/bin/env python3

import base64
import json
import os
import pathlib

def email_preview(dir = os.getcwd()):
    dir = pathlib.Path(dir)

    json_file = dir / ".output_metadata.json"
    with open(json_file, "r") as f:
        payload = json.load(f)

    email_dir = dir / "email-preview"
    email_dir.mkdir(exist_ok = True)

    html = payload.get("rsc_email_body_html")
    html = html.replace('src="cid:', 'src="')

    html_file = email_dir / "index.html"
    with open(html_file, "w") as f:
        f.write(html)

    images = payload.get("rsc_email_images", {})
    if images:
        # https://github.com/rich-iannone/quarto-email/issues/4
        for name, encoded in images.items():
            decoded = base64.b64decode(encoded)
            image_file = email_dir / name
            with open(image_file, "wb") as f:
                f.write(decoded)

    return html_file

if __name__ == '__main__':
    import sys
    dir = os.getcwd()
    if len(sys.argv) > 1:
        dir = sys.argv[1]
    print(email_preview(dir))
```

**Comment by rich-iannone** [comment]
@aronatkins These are nice developer-facing functions, wish I had them sooner! If we make this public, where should they live? Good options might be rsconnect (R and Python variants). Maybe in R blastula and in a future Python blastula. This could be also be packaged along with the shortcode extension, perhaps within the main README or a separate one? And maybe separate .R and .py files could also be included.

**Comment by aronatkins** [comment]
Folks should be able to preview their email messages without using an external tool. I don't know how that is best accomplished with Quarto / this plugin.

Thinking about ergonomics a little, wouldn't it be nice if:

1. `quarto preview` launches two browser sessions: one previewing the document, one previewing the email.
2. the email HTML is dynamically constructed on-the-fly by the plugin and served.

The other option would be to bake the capabilities of these scripts into the plugin - and have it enabled if some option / environment variable is set. Render an email directory alongside the report HTML and directory.

**Comment by rich-iannone** [comment]
Aron thanks for the advice here. I chose the less ideal option (render email HTML file in a subdirectory) only because the better option is currently difficult to implement with this being a standalone extension. This will change in the future though, and we’ll provide the preview in a second revision.

How to generate the email preview document is now documented in the README. I did a number of tests and found that the subdirectory approach is ideal (doesn’t interfere during deployment to connect).
